### PR TITLE
Add  com.github.babluboy.nutty

### PR DESCRIPTION
**Pull request by babluboy**
Initial Flatpak request for Nutty - A simple application to view network details

### CODE REVIEW

File: com.github.babluboy.nutty.json [hunk 1]
```diff
@@ -0,0 +1,93 @@
+{
+    "id": "com.github.babluboy.nutty",
+    "runtime": "org.gnome.Platform",
+    "runtime-version" : "3.24",
+    "sdk": "org.gnome.Sdk",
+    "command": "bookworm",
+    "rename-appdata-file": "com.github.babluboy.nutty.appdata.xml",
```
Comment (bilelmoussaoui) [inline com.github.babluboy.nutty.json]: Wouldn't this be done upstream too?

File: com.github.babluboy.nutty.json [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+					"sha256": "42fe6d651910cb8b720167f71c5255a1b7b1afc82fecd3f31e61f9602b3b1335"
+				}
+			]
+		},
+		{
+			"name": "granite",
+			"buildsystem": "cmake",
+			"config-opts": [
+				"-DCMAKE_BUILD_TYPE=Release",
+				"-DCMAKE_INSTALL_PREFIX=/app",
```
Comment (bilelmoussaoui) [inline com.github.babluboy.nutty.json]: You don't need to specify the prefix nor the libdir anymore

File: com.github.babluboy.nutty.json [hunk 3]
```diff
@@ -0,0 +1,93 @@
+{
+    "id": "com.github.babluboy.nutty",
+    "runtime": "org.gnome.Platform",
+    "runtime-version" : "3.24",
```
Comment (bilelmoussaoui) [inline com.github.babluboy.nutty.json]: Won't the application run on newer runtimes? 3.26 for example?

Comment (TingPing) [reply com.github.babluboy.nutty.json]: There is little reason to use 3.24 over 3.28

File: com.github.babluboy.nutty.json [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    "sdk": "org.gnome.Sdk",
+    "command": "nutty",
+    "rename-appdata-file": "com.github.babluboy.nutty.appdata.xml",
```
Comment (TingPing) [inline com.github.babluboy.nutty.json]: It is already correct, so renaming it doesn't make sense.

File: com.github.babluboy.nutty.json [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    ],
+
+    "build-options": {
+        "cflags": "-O2",
+        "cxxflags": "-O2"
+    },
+    "modules": [
+         {
+                "name": "openjpeg",
+                "buildsystem" : "cmake",
```
Comment (TingPing) [inline com.github.babluboy.nutty.json]: `cmake-ninja` and add `"config-opts": ["-DCMAKE_BUILD_TYPE=Release"]`, same for all cmake projects.

File: com.github.babluboy.nutty.json [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+					"sha256": "035a9b36ef73352ff8a76be06b876d9e2ea6633abb693fb67297d998c8d3a240"
+				}
+			]
+        },
+        {
+            "name": "nutty",
+            "buildsystem": "cmake",
+            "build-options" : {
+                    "config-opts" : [
+                            "-DCMAKE_INSTALL_PREFIX=/app"
```
Comment (TingPing) [inline com.github.babluboy.nutty.json]: This doesn't need to be added.

File: com.github.babluboy.nutty.json [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+                    ]
+            },
+            "sources": [
+                {
+                    "type": "archive",
+                    "url": "https://github.com/babluboy/nutty/archive/1.0.0.tar.gz",
+                    "sha256": "6a6babeb530f49a886174b7cec7a00f47ab73cce2ad680bd7dd1018b1e67f860",
+                    "branch": "1.0.0",
```
Comment (TingPing) [inline com.github.babluboy.nutty.json]: You left git properties here.

**Comment by TingPing** [comment]
bot, build com.github.babluboy.nutty

**Comment by babluboy** [comment]
@TingPing @nedrichards 
Looks like the build for nutty failed with the following:
```
-- Checking for modules 'gtk+-3.0;gee-0.8;granite>=0.5;sqlite3>=3.5.9;libxml-2.0;libnotify'
--   
CMake Error at /usr/share/cmake-3.7/Modules/FindPkgConfig.cmake:415 (message):
  A required package was not found

```
https://flathub.org/builds/#/builders/40/builds/82
Any help on what might be wrong, this builds fine with Launchpad, Elementary Appcenter and on my local.

**Comment by TingPing** [comment]
libnotifiy isn't in the sdk so you need to build it too.

**Comment by babluboy** [comment]
@TingPing Thanks for the update. I have now added the source for libnotify in the json. Can you re-start the build, so that we can see if nutty builds fine now. Thanks.

**Comment by TingPing** [comment]
bot, build com.github.babluboy.nutty

**Comment by babluboy** [comment]
@TingPing Looks like the build is still failing even after adding the libnotify source in the json manifest. Which one i'm i missing from this list: gtk+-3.0;gee-0.8;granite>=0.5;sqlite3>=3.5.9;libxml-2.0;libnotify

Do I have to add sources for sqlite3 and libxml as well ?

File: com.github.babluboy.nutty.json [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+			"build-options": {
+				"env": {
+					"PKG_CONFIG_GOBJECT_INTROSPECTION_1_0_GIRDIR": "/app/share/gir-1.0",
+					"PKG_CONFIG_GOBJECT_INTROSPECTION_1_0_TYPELIBDIR": "/app/lib/girepository-1.0"
+				}
+			},
+			"sources": [
+				{
+					"type":"archive",
+					"url": "https://github.com/elementary/granite/archive/0.4.1.tar.gz",
```
Comment (bilelmoussaoui) [inline com.github.babluboy.nutty.json]: The application do requires granite 0.5 but you're building 0.4.1 here

File: com.github.babluboy.nutty.json (hunk 4 above)
Comment (babluboy) [reply com.github.babluboy.nutty.json]: @TingPing Thanks, removed the rename clause

File: com.github.babluboy.nutty.json (hunk 3 above)
Comment (babluboy) [reply com.github.babluboy.nutty.json]: @TingPing thanks, updated to 3.28

**Comment by nedrichards** [comment]
bot, build com.github.babluboy.nutty

**Comment by babluboy** [comment]
Oops - the build progressed but failed with the appdata parsing:
```
Processing application com.github.babluboy.nutty
Error loading AppData file: failed to parse /app/share/appdata/com.github.babluboy.nutty.appdata.xml: Error on line 16 char 16: <ul> already set '​' and tried to replace with '
​           '
Error: ERROR: appstream-compose failed: Child process exited with code 1
```
Here is the appdata section where it is complaining:
      ```
<ul>
​           <li>My Info: Provides basic and detailed information for the device network card</li>
​           <li>Usage: Provides network data usage in two views - historical usage and current usage</li>
           <li>Speed: Check Upload and Download speeds and get route times to a host</li>
           <li>Ports: Provides information on active ports and application using them on the local device</li>
           <li>Devices: Monitors, alerts and provides information on the other devices on the network</li>
​       </ul>
```
Is it due to the "-" character in ".. in two views - historical usage..." ?

**Comment by Alexander-Wilms** [comment]
I think I once had the same issue with an appdata file. It contained some unusual whitespace characters.

**Comment by babluboy** [comment]
@Alexander-Wilms Thanks for the update, any suggestions on how I can find such characters in my appdatafile or is there a utility which I can run to detect the same or probably a local version of the script that flatpak uses to see what is causing the issue

**Comment by babluboy** [comment]
Googled a bit to find from [here](https://github.com/bitseater/weather/issues/38) that flatpak appstream uses "appstream-util" and was able to replicate the issue by running appstream-util on the appdata.xml for nutty. Hopefully I should be able to resolve this by playing around with the appdata contents.

**Comment by babluboy** [comment]
Looks like appstream-util is a lot stricter than appstreamcli. I have got the issues with the tags fixed but looks like it needs me to ensure that the screen shot images are in 16:9 ratio. Will fix the same and create a release soon.

Just noting here an online tool for cropping images quickly in the 16:9 ratio: https://ezgif.com/crop

**Comment by babluboy** [comment]
@TingPing @nedrichards Made the changes to appdata to validate correctly by appstream-util. Created a new release and updated the json manifest.
Can you please schedule a flatpak build.

**Comment by nedrichards** [comment]
bot, build com.github.babluboy.nutty

**Comment by nedrichards** [comment]
> Error: Command 'nutty' not found

https://flathub.org/builds/#/builders/2/builds/1101

**Comment by babluboy** [comment]
@nedrichards @TingPing  Many thanks for your support. Have corrected the command to mention the full RDNN: `"command": "com.github.babluboy.nutty"`

Hopefully the build should progress a little more. Can you please schedule another build for Nutty.

File: com.github.babluboy.nutty.json [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    "runtime": "org.gnome.Platform",
+    "runtime-version" : "3.28",
+    "sdk": "org.gnome.Sdk",
+    "command": "com.github.babluboy.nutty",
+    "finish-args": [
+            "--share=ipc",
+            "--socket=x11",
+            "--device=dri",
+            "--socket=wayland",
+            "--filesystem=home",
```
Comment (bilelmoussaoui) [inline com.github.babluboy.nutty.json]: As it's a network utility application, how does it not require the network access?

**Comment by nedrichards** [comment]
bot, build com.github.babluboy.nutty

**Comment by nedrichards** [comment]
I was interested in the app so I pulled it and had a bit of a hack. At the very least you need pciutils here or the app crashes on startup. You must test the app before you submit it here or you're wasting users and reviewers time. The app launches but there's still a bunch of errors and I'm not entirely sure it's going to do anything:

> [WARNING 18:51:32.530177] nutty.vala:1147: Failure in executing async command [pkexec /app/share/nutty/scripts/nutty_nethogs_script.sh Select a connection] : Failed to execute child process “pkexec” (No such file or directory)
[WARNING 18:51:42.190972] nutty.vala:1147: Failure in executing async command [pkexec /app/share/nutty/scripts/nutty_nethogs_script.sh Select a connection] : Failed to execute child process “pkexec” (No such file or directory)
[WARNING 18:51:42.742161] nutty.vala:1147: Failure in executing async command [pkexec /app/share/nutty/scripts/nutty_nethogs_script.sh Select a connection] : Failed to execute child process “pkexec” (No such file or directory)
[WARNING 18:51:45.994159] nutty.vala:1147: Failure in executing async command [netstat -p -e] : Failed to execute child process “netstat” (No such file or directory)

I can see a way towards bundling net-tools to get netstat but `pkexec` just isn't going to run under flatpak as far as I can see. Unless someone else knows more I'm not sure that this app can be distributed in a Flatpak without a refactor.

**Comment by nedrichards** [comment]
bot, build com.github.babluboy.nutty

File: com.github.babluboy.nutty.json (hunk 9 above)
Comment (TingPing) [reply com.github.babluboy.nutty.json]: `--share=network`

**Comment by babluboy** [comment]
@nedrichards Many thanks for your updates and checks. Apologies, I'm not sure how I could have tested the app on Flatpak before going through this process (will be good to know for the future if there is a local way to test this prior to submitting). The app works fine on Elementary OS and Ubuntu, so I thought it will be good to distribute as Flatpak.

If we cannot have `pkexec` in Flatpak then all the places where the user needs sudo will not work - if there is no further way to achieve this then we need not proceed to have this app in Flatpak. If `pkexec` can be programaitically achieved via a library or api then I can implement in code in a later version.

Happy to get your further thoughts on this.

**Comment by bilelmoussaoui** [comment]
@nedrichards There's also `flatpak-builder install` https://github.com/flatpak/flatpak-docs/blob/master/docs/flatpak-builder.rst#installing-builds-directly

**Comment by nedrichards** [comment]
@bilelmoussaoui absolutely. as a developer that's likely easier - I sometimes like to look at the repo contents, but that shouldn't be necessary for normal people.

**Comment by babluboy** [comment]
@bilelmoussaoui  @nedrichards Many thanks for your suggestions on how to test locally. I will try that out. If I find a way forward in programatically implementing the pkexec functionality, then I will come back here and ask you to progress nutty as flatpak. 

I have been able to build this as a docker image, so that can be an alternative until a flatpak is possible.

**Comment by barthalion** [comment]
I'm closing it due to inactivity – but please re-open it or resubmit as new pull request if you would like to see it merged in the future.

**Pull request closed without merging**